---
title: "PAMES"
author: "Dario Romagnoli"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{PAMES}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

**PAMES** (*Purity Assessment from MEthylation Sites*) is a tool developed 
for estimating purity of tumor samples.
The default process for data obtained with Illumina BeadChip 
(450k or 27k) is characterized by two main steps:

1. computation of AUC (`compute_AUC`)
2. selection of informative CpG __sites__ (`select_informative_sites`)
3. estimation of purity (`compute_purity`)

For data generated by Bisulphite Sequencing (BS) technology, the purity estimation 
requires two previous additional steps:
  
1. computing indexes mapping CpG __sites__ to to CpG __islands__ (`compute_island_indexes`)
2. reducing the data to CpG islands (`reduce_to_islands`)

Those steps convert beta values from different CpG sites 
to a single beta value associated to the CpG island which CpG sites belong 
to (CpG sites located outsite CpG islands are discared).

After that, one can proceed as previously described (beware the different *select* function):
1. computation of AUC (`compute_AUC`)
2. selection of informative CpG __islands__ (`select_informative_islands`)
3. estimation of purity (`compute_purity`)

A set of pre-computed CpG sites (for Illumina 450k) and CpG islands for 14 cancer
types are available with our [PAMESdata](https://github.com/cgplab/PAMESdata) package.

## Examples

Below are showed examples of the steps required to run analysis 
with different data types.

### Illumina 450k (hg19) -- pre-computed informative CpG sites

    purity_brca <- compute_purity(BRCA_matrix, BRCA_sites)
    
### Illumina 27k (hg38) -- compute new set of informative CpG sites

    auc_scores <- compute_AUC(tumor_matrix, control_matrix)
    cpg_sites <- select_informative_sites(tumor_matrix, control_matrix, 
      auc_scores, genome="hg38")
    tumor_purity <- compute_purity(tumor_matrix, cpg_sites)

### Bisulphite Sequencing -- pre-computed informative CpG islands

    bs_BRCA_idx <- compute_island_indexes(bs_BRCA_coord_table)
    bs_BRCA_reduced <- reduce_to_islands(bs_BRCA_matrix, bs_BRCA_idx)
    bs_BRCA_purity <- compute_purity(bs_BRCA_reduced, BRCA_islands)
    
### Bisulphite Sequencing -- compute new set of informative CpG islands

Tip: `reduce_to_islands` is a time consuming step (until we implement a parallel
approch, asap) so it's preferable to perform it on one single matrix 
with tumor and normal samples together and then split the matrix (see example below).

    |           | tumor1| tumor2| control1| control2|
    |:----------|------:|------:|--------:|--------:|
    |chr1-10470 |   0.98|   0.60|     0.80|     0.06|
    |chr1-10484 |   0.55|   0.24|     0.26|     0.56|
    |.......... |   ....|   ....|     ....|     ....|
    |chr2-10555 |   0.60|   0.04|     0.10|     1.00|
    |chr2-10610 |   0.69|   0.64|     0.57|     0.80|
    

    bs_BRCA_idx <- compute_island_indexes(bs_BRCA)
    bs_BRCA_reduced <- reduce_to_islands(bs_BRCA, bs_BRCA_idx)
    bs_BRCA_redux <- reduce_bs_data(bs_brca, bs_)
    bs_BRCA_tumor <- bs_BRCA_redux[, 1:2]
    bs_BRCA_control <- bs_BRCA_redux[, 3:4]
    bs_BRCA_islands <- select_informative_islands(bs_BRCA_tumor, bs_BRCA_control)
    bs_BRCA_purity <- compute_purity(bs_BRCA_tumor, bs_BRCA_islands)
