---
title: "PAMES"
author: "Dario Romagnoli"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{PAMES}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

**PAMES** (*Purity Assessment from MEthylation Sites*) is a tool developed 
for estimating purity of tumor samples.
The default process for data obtained with Illumina BeadChip 
(450k or 27k) is divided into three steps:

1. computation of AUC (`compute_AUC`)
2. selection of informative CpG __sites__ (`select_informative_sites`)
3. estimation of purity (`compute_purity`)

For data generated by Bisulphite Sequencing (BS) technology, the purity estimation 
requires two preceding additional steps:
  
1. computing indexes mapping CpG __sites__ to to CpG __islands__ (`compute_island_indexes`)
2. reducing the data to CpG islands (`reduce_to_islands`)

These steps convert beta values from different CpG sites 
to a single beta value associated to the CpG island which CpG sites belong 
to (CpG sites located outsite CpG islands are discared).

Next, the analysis proceedes as previously described (beware the different *select* function):

1. computation of AUC (`compute_AUC`)
2. selection of informative CpG __islands__ (`select_informative_islands`)
3. estimation of purity (`compute_purity`)

A set of pre-computed CpG sites (for Illumina 450k) and CpG islands for 14 cancer
types are available with our [PAMESdata](https://github.com/cgplab/PAMESdata) package.

## Examples

Below are showed examples of the steps required to run analyses 
with different data types.

### Illumina BeadChip
#### Pre-computed informative CpG sites (BRCA_sites) -- 450k, hg19

    BRCA_purity <- compute_purity(BRCA_tumor, BRCA_sites)
    
#### New set of informative CpG sites -- 27k, hg38

    auc_scores <- compute_AUC(tumor_matrix, control_matrix)
    tumor_sites <- select_informative_sites(tumor_matrix, control_matrix, auc_scores, genome="hg38")
    tumor_purity <- compute_purity(tumor_matrix, cpg_sites)

### Bisulphite Sequencing
#### Data structure

    bs_coordinates     bs_matrix
    |chr  |  coord|    | tumor1| tumor2| control1| control2|
    |:----|------:|    |------:|------:|--------:|--------:|
    |chr1 | 135207|    |   0.98|   0.60|     0.80|     0.06|
    |chr1 | 138290|    |   0.55|   0.24|     0.26|     0.56|
    |.... | ......|    |   ....|   ....|     ....|     ....|
    |chr2 | 661905|    |   0.60|   0.04|     0.10|     1.00|
    |chr2 | 661925|    |   0.69|   0.64|     0.57|     0.80|

### Pre-computed informative CpG islands (BRCA_islands)
 
    bs_BRCA_idx <- compute_islands_indexes(bs_BRCA_coordinates)
    bs_BRCA_reduced <- reduce_to_islands(bs_BRCA_tumor, bs_BRCA_idx)
    bs_BRCA_purity <- compute_purity(bs_BRCA_tumor, BRCA_islands)
    
### New set of informative CpG islands

    bs_idx <- compute_island_indexes(bs_coordinates)
    bs_reduced <- reduce_to_islands(bs_matrix, bs_idx)
    bs_tumor_matrix <- bs_reduced[, 1:2]                                         # <<< ADDITIONAL STEP
    bs_control_matrix <- bs_reduced[, 3:4]                                       # <<< ADDITIONAL STEP
    bs_islands <- select_informative_islands(bs_tumor_matrix, bs_control_matrix) # <<< ADDITIONAL STEP
    bs_purity <- compute_purity(bs_tumor_matrix, bs_islands)
    
**NOTE**

- `reduce_to_islands` is the bottleneck of the analysis 
(until we implement a parallel approch, asap):
it's preferable to perform it on one single matrix 
with tumor and normal samples together and then split the matrix (see example above).
